pg_shard.c

```diff
@@ -807,11 +1081,135 @@ BuildLocalQuery(Query *query, List *localRestrictList)
 	Assert(list_length(joinTree->fromlist) == 1);
 	joinTree->quals = (Node *) make_ands_explicit((List *) localRestrictList);
 
+	if (safeToPushDownAggregate)
+	{
+		localQuery = RemoveAggregates(localQuery);
+	}
+
 	return localQuery;
 }
 
 
 /*
+ * RemoveAggregates gets a query which includes aggregations and returns a non-aggregated
+ * query. It converts all range table entries (including AggRefs) to corresponding Vars.
+ * Also, this function eliminates GROUP BY/ORDER BY columns that appears on range table
+ * entries.
+ */
+static Query *
+RemoveAggregates(Query *aggregatedQuery)
+{
+	Query *nonAggregatedQuery = copyObject(aggregatedQuery);
+	Index masterTableId = 1;
+	AttrNumber columnId = 1;
+	List *aggregatedTargetList = aggregatedQuery->targetList;
+	ListCell *targetListCell = NULL;
+	List *nonAggregatedTargetList = NIL;
+	AttrNumber targetResNo = 1;
+
+	foreach(targetListCell, aggregatedTargetList)
+	{
+		TargetEntry *targetListEntry = (TargetEntry *) lfirst(targetListCell);
+		TargetEntry *newTargetEntry = copyObject(targetListEntry);
+
+		/* update target list entries which include Aggrefs */
+		if (ContainsAggref((Node *) newTargetEntry->expr))
+		{
+			Var *targetVar = makeVarFromTargetEntry(masterTableId, targetListEntry);
+
+			newTargetEntry = makeTargetEntry((Expr *) targetVar,
+											 targetResNo,
+											 targetListEntry->resname,
+											 targetListEntry->resjunk);
+		}
+
+		/* update column attribute numbers */
+		newTargetEntry->expr = (Expr *) VarNodeMutator((Node *) newTargetEntry->expr,
+													   &columnId);
+
+		/* ressortgroupref and resjunk must be updated for ORDER BY columns */
+		newTargetEntry->ressortgroupref = targetListEntry->ressortgroupref;
+		newTargetEntry->resjunk = targetListEntry->resjunk;
+
+		newTargetEntry->resno = targetResNo;
+		nonAggregatedTargetList = lappend(nonAggregatedTargetList,
+										  newTargetEntry);
+		++targetResNo;
+	}
+
+	nonAggregatedQuery->targetList = list_copy(nonAggregatedTargetList);
+	nonAggregatedQuery->hasAggs = false;
+	nonAggregatedQuery->groupClause = NIL;
+	nonAggregatedQuery->havingQual = NULL;
+
+	return nonAggregatedQuery;
+}
+
+
+/*
+ * VarNodeMutator walks over the original expression,  and recurses into
+ * the expression mutator function for all other expression types to update the
+ * varattno of the Vars in the expression.
+ */
+static Node *
+VarNodeMutator(Node *originalNode, AttrNumber *columnId)
+{
+	Node *newNode = NULL;
+	if (originalNode == NULL)
+	{
+		return NULL;
+	}
+
+	if (IsA(originalNode, Var))
+	{
+		uint32 masterTableId = 1; /* one table on the master node */
+		Var *newColumn = copyObject(originalNode);
+		newColumn->varno = masterTableId;
+		newColumn->varattno = (*columnId);
+		(*columnId)++;
```

Comment:
I don't want to explain it here.
I updated the function name and comment, let's see if it is easier to understand now. If still not, I can explain what the function does in detail, and we can decide on the name and comments .